core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -63,80 +66,136 @@ public class SharePartitionManager {
     private final ReplicaManager replicaManager;
     private final Time time;
     private final ShareSessionCache cache;
+    private ConcurrentLinkedQueue<ShareFetchPartitionData> fetchQueue;
+    private final Lock lock;
 
     public SharePartitionManager(ReplicaManager replicaManager, Time time, ShareSessionCache cache) {
         this.replicaManager = replicaManager;
         this.time = time;
         this.cache = cache;
         partitionCacheMap = new ConcurrentHashMap<>();
+        fetchQueue = new ConcurrentLinkedQueue<>();
+        this.lock = new ReentrantLock();
     }
 
     // TODO: Move some part in share session context and change method signature to accept share
     //  partition data along TopicIdPartition.
     public CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>> fetchMessages(
-        String groupId,
-        String memberId,
-        FetchParams fetchParams,
-        List<TopicIdPartition> topicIdPartitions) {
+            String groupId,
+            String memberId,
+            FetchParams fetchParams,
+            List<TopicIdPartition> topicIdPartitions) {
         log.trace("Fetch request for topicIdPartitions: {} with groupId: {} fetch params: {}",
-            topicIdPartitions, groupId, fetchParams);
+                topicIdPartitions, groupId, fetchParams);
         CompletableFuture<Map<TopicIdPartition, ShareFetchResponseData.PartitionData>> future = new CompletableFuture<>();
-
-        synchronized (this) {
-            Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = new HashMap<>();
-            topicIdPartitions.forEach(topicIdPartition -> {
-                // TODO: Fetch inflight and delivery count from config.
-                SharePartition sharePartition = partitionCacheMap.computeIfAbsent(sharePartitionKey(groupId, topicIdPartition),
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = new HashMap<>();
+        topicIdPartitions.forEach(topicIdPartition -> {
+            // TODO: Fetch inflight and delivery count from config.
+            SharePartition sharePartition = partitionCacheMap.computeIfAbsent(sharePartitionKey(groupId, topicIdPartition),
                     k -> new SharePartition(groupId, topicIdPartition, 100, 5));
-                topicPartitionData.put(topicIdPartition, new FetchRequest.PartitionData(
-                    topicIdPartition.topicId(),
-                    sharePartition.nextFetchOffset(),
-                    -1,
-                    Integer.MAX_VALUE,
-                    Optional.empty()));
-            });
+            lock.lock();
+            try {
+                // we add a share partition to the request for fetch queue only if its fetch lock is available
+                if (sharePartition.isFetchLockAvailable()) {
```

Comment:
Rather than calling `isFetchLockAvailable` we should directly try acquiring, which ever request acquires the share partition lock gets the access.

Reply:
Moreover I think we should move lock acquiring capability to `fetchResults` method itself, the request can be send as it is in queue.